src/hotspot/share/services/threadService.cpp

```diff
@@ -886,6 +886,9 @@ void ThreadSnapshot::initialize(ThreadsList * t_list, JavaThread* thread) {
       _thread_status == java_lang_Thread::IN_OBJECT_WAIT_TIMED) {
 
     Handle obj = ThreadService::get_current_contended_monitor(thread);
```

Comment:
There must be a safepoint here then.
I think this would be better and safer if blocker_object and blocker_object_owner are Handles.  Can you change them to Handles?

Reply:
I can't see anywhere a safepoint check would occur in that code. This issue was flagged as being in Loom so perhaps the loom code is different and is what introduces the safepoint check?
But I agree with Coleen that the best solution is to just use Handles.

Reply:
It is not loom-specific and reproduced n jdk/jdk with -XX:+CheckUnhandledOops.

Reply:
What do you think about moving 
    Handle obj = ThreadService::get_current_contended_monitor(thread);
out of scope of block_object oop visibility? 
It is my second patch.

Reply:
I'm missing something. How can a NULL oop get corrupted even if there is a GC?

Reply:
This is a specific of "CheckUnhandledOops"
I've written in bug comment "Another possible fix would be to disable corruption of NULL unhandled oops. They couldn't be changed really."

We discussed it with Coleen and seems that moving NULL oops out of possible safepoint or handling them seems easier option than changing UnhandledOops.cpp to don't corrupt NULL. It is here:
https://github.com/openjdk/jdk/blob/77bdc3065057b07a676b010562c89bb0f21512b7/src/hotspot/share/runtime/unhandledOops.cpp#L113

Reply:
ThreadService::get_current_contended_monitor calls Thread::check_for_dangling_thread_pointer calls          ThreadsSMRSupport::is_a_protected_JavaThread_with_lock((JavaThread *) thread),

The potential safepoint is here, where CheckUnhandledOops puts junk in any oop on the stack.

inline bool ThreadsSMRSupport::is_a_protected_JavaThread_with_lock(JavaThread *thread) {
  MutexLocker ml(Threads_lock->owned_by_self() ? NULL : Threads_lock);
  return is_a_protected_JavaThread(thread);
}

Reply:
Thanks Coleen. I'm still not sure that CheckUnhandledOops should be touching NULL oops but ...

Leonid the workaround seems okay.